### ISSUE TITLE
Test failure: System.Net.Security.Tests.ServerAsyncAuthenticateTest.ServerAsyncAuthenticate_MismatchProtocols_Fails(serverProtocol: Tls12, clientProtocol: Tls, expectedException: typeof(System.Security.Authentication.AuthenticationException))

### ISSUE THREAD

**Issue by v-haren** [dotnet/runtime#36192]
failed in job: [runtime 20200510.20 ](https://dev.azure.com/dnceng/public/_build/results?buildId=638932&view=ms.vss-test-web.build-test-results-tab&runId=19860426&resultId=150684&paneView=debug)

Error message
~~~
Assert.IsAssignableFrom() Failure
Expected: typeof(System.Security.Authentication.AuthenticationException)
Actual: typeof(System.TimeoutException)


Stack trace
   at System.Net.Security.Tests.ServerAsyncAuthenticateTest.ServerAsyncAuthenticate_MismatchProtocols_Fails(SslProtocols serverProtocol, SslProtocols clientProtocol, Type expectedException) in /_/src/libraries/System.Net.Security/tests/FunctionalTests/ServerAsyncAuthenticateTest.cs:line 62
--- End of stack trace from previous location ---
~~~

**Comment by wfurt** [comment]
This is curious. Ubuntu.1604.Amd64.Open.

```
    System.Net.Security.Tests.ServerAsyncAuthenticateTest.ServerAsyncAuthenticate_MismatchProtocols_Fails(serverProtocol: Tls12, clientProtocol: Tls, expectedException: typeof(System.Security.Authentication.AuthenticationException)) [FAIL]
      Assert.IsAssignableFrom() Failure
      Expected: typeof(System.Security.Authentication.AuthenticationException)
      Actual:   typeof(System.TimeoutException)
      Stack Trace:
        /_/src/libraries/System.Net.Security/tests/FunctionalTests/ServerAsyncAuthenticateTest.cs(62,0): at System.Net.Security.Tests.ServerAsyncAuthenticateTest.ServerAsyncAuthenticate_MismatchProtocols_Fails(SslProtocols serverProtocol, SslProtocols clientProtocol, Type expectedException)
        --- End of stack trace from previous location ---
      Output:
        Server: Tls; Client: Tls12 expectedToFail: True
        Client exception: System.Security.Authentication.AuthenticationException: Authentication failed, see inner exception.
         ---> Interop+OpenSsl+SslException: SSL Handshake failed with OpenSSL error - SSL_ERROR_SSL.
         ---> Interop+Crypto+OpenSslCryptographicException: error:14077102:SSL routines:SSL23_GET_SERVER_HELLO:unsupported protocol
           --- End of inner exception stack trace ---
           at Interop.OpenSsl.DoSslHandshake(SafeSslHandle context, ReadOnlySpan`1 input, Byte[]& sendBuf, Int32& sendCount) in /_/src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs:line 314
           at System.Net.Security.SslStreamPal.HandshakeInternal(SafeFreeCredentials credential, SafeDeleteSslContext& context, ReadOnlySpan`1 inputBuffer, Byte[]& outputBuffer, SslAuthenticationOptions sslAuthenticationOptions) in /_/src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Unix.cs:line 115
           --- End of inner exception stack trace ---
           at System.Net.Security.SslStream.ForceAuthenticationAsync[TIOAdapter](TIOAdapter adapter, Boolean receiveFirst, Byte[] reAuthenticationData, Boolean isApm) in /_/src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs:line 297
           at System.Threading.Tasks.TaskTimeoutExtensions.TimeoutAfter(Task task, TimeSpan timeout) in /_/src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs:line 38
           at System.Net.Security.Tests.ServerAsyncAuthenticateTest.ServerAsyncSslHelper(SslProtocols clientSslProtocols, SslProtocols serverSslProtocols, Boolean expectedToFail) in /_/src/libraries/System.Net.Security/tests/FunctionalTests/ServerAsyncAuthenticateTest.cs:line 159
  Finished:    System.Net.Security.Tests

```

It does not look like timing since the client throws `unsupported protocol`

cc: @safern

**Comment by wfurt** [comment]
I was able to collect some data with strace. (215 is client, 219 is the server) 
```
[pid 29583] 17:55:42.752313 connect(215, {sa_family=AF_INET6, sin6_port=htons(46805), inet_pton(AF_INET6, "::ffff:127.0.0.1", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=0}, 28) = -1 EINPROGRESS (Operation now in progress)
[pid 29586] 17:55:42.757499 getsockname(219, {sa_family=AF_INET, sin_port=htons(46805), sin_addr=inet_addr("127.0.0.1")}, [16]) = 0
[pid 29586] 17:55:42.757588 getsockname(215, {sa_family=AF_INET6, sin6_port=htons(58206), inet_pton(AF_INET6, "::ffff:127.0.0.1", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=0}, [28]) = 0
[pid 29586] 17:55:42.759323 getsockname(215, {sa_family=AF_INET6, sin6_port=htons(58206), inet_pton(AF_INET6, "::ffff:127.0.0.1", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=0}, [28]) = 0
[pid 29586] 17:55:42.759390 getsockname(219, {sa_family=AF_INET, sin_port=htons(46805), sin_addr=inet_addr("127.0.0.1")}, [16]) = 0
[pid 29586] 17:55:42.759948 sendmsg(215, {msg_name(0)=NULL, msg_iov(1)=[{"\26\3\1\1\6\1\0\1\2\3\2E\300-\345\2?\236H\342\342k\242\36\2236\vc\301Xa\355\224\365\33\307\35064\267\3539\314\0\0b\300\24\300\n\0009\0008\0007\0006\0\210\0\207\0\206\0\205\300\17\300\5\0005\0\204\300\23\300\t\0003\0002\0001\0000\0\232\0\231\0\230\0\227\0E\0D\0C"..., 267}], msg_controllen=0, msg_flags=0}, 0) = 267
[pid 29586] 17:55:42.760070 recvmsg(215, 0x7f77b1ff87d8, 0) = -1 EAGAIN (Resource temporarily unavailable)
[pid 29586] 17:55:42.760346 fcntl(219, F_GETFL) = 0x2 (flags O_RDWR)
[pid 29586] 17:55:42.760383 fcntl(219, F_SETFL, O_RDWR|O_NONBLOCK) = 0
[pid 29586] 17:55:42.760440 recvmsg(219, {msg_name(0)=NULL, msg_iov(1)=[{"\26\3\1\1\6\1\0\1\2\3\2E\300-\345\2?\236H\342\342k\242\36\2236\vc\301Xa\355\224\365\33\307\35064\267\3539\314\0\0b\300\24\300\n\0009\0008\0007\0006\0\210\0\207\0\206\0\205\300\17\300\5\0005\0\204\300\23\300\t\0003\0002\0001\0000\0\232\0\231\0\230\0\227\0E\0D\0C"..., 4128}], msg_controllen=0, msg_flags=0}, 0) = 267
[pid 29586] 17:55:42.765258 sendmsg(219, {msg_name(0)=NULL, msg_iov(1)=[{"\26\3\1\0B\2\0\0>\3\1\t\347-\267\273\243]*}\221>j,}\254\276G#P:\223\23\302R\357\270\263EN\20\25C\0\300\24\0\0\26\377\1\0\1\0\0\v\0\4\3\0\1\2\0#\0\0\0\17\0\1\1\26\3\1\5\207\v\0\5\203\0\5\200\0\5}0\202\5y0\202\4a\240\3\2\1\2\2"..., 2102}], msg_controllen=0, msg_flags=0}, 0 <unfinished ...>
[pid 29586] 17:55:42.765523 recvmsg(219, 0x7f77b1ff87d8, 0) = -1 EAGAIN (Resource temporarily unavailable)
[pid 29576] 17:55:42.780885 recvmsg(215, {msg_name(0)=NULL, msg_iov(1)=[{"\26\3\1\0B\2\0\0>\3\1\t\347-\267\273\243]*}\221>j,}\254\276G#P:\223\23\302R\357\270\263EN\20\25C\0\300\24\0\0\26\377\1\0\1\0\0\v\0\4\3\0\1\2\0#\0\0\0\17\0\1\1\26\3\1\5\207\v\0\5\203\0\5\200\0\5}0\202\5y0\202\4a\240\3\2\1\2\2"..., 4128}], msg_controllen=0, msg_flags=0}, 0) = 2102
[pid 29576] 17:55:42.781285 sendmsg(215, {msg_name(0)=NULL, msg_iov(1)=[{"\25\3\1\0\2\2F", 7}], msg_controllen=0, msg_flags=0}, 0 <unfinished ...>
[pid 29586] 17:55:45.773170 close(215)  = 0
[pid 29586] 17:55:45.773444 close(219)  = 0
```
The client sends 7 byte alert but the is not received on server even if the read is pending. 
I was suspecting Nagle's algorithm bet setting NoDelay to true on socket did not fix it. 

I also have somewhat improved trace. The alert message is received after we start closing downs. 
```
<Event MSec=  "9515.6585" PID="24252" PName="Process(24252)" TID="24248" EventName="SslStreamCtor" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#22286816" localId="127.0.0.1:42313" remoteId="127.0.0.1:33734"/>
<Event MSec=  "9515.8923" PID="24252" PName="Process(24252)" TID="24248" EventName="SslStreamCtor" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#40350611" localId="[::ffff:127.0.0.1]:33734" remoteId="[::ffff:127.0.0.1]:42313"/>
<Event MSec=  "9519.4433" PID="24252" PName="Process(24252)" TID="24248" EventName="SecureChannelCtor" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SecureChannel#58207618" sslStream="SslStream#40350611" hostname="testservereku.contoso.com" clientCertificatesCount="0" encryptionPolicy="0"/>
<Event MSec=  "9520.0616" PID="24252" PName="Process(24252)" TID="24248" EventName="SentFrame" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#40350611" frameInfo="Tls:ClientHello[334] TargetName=&apos;testservereku.contoso.com&apos; SupportedVersion=&apos;Tls, Tls12&apos; ApplicationProtocols=&apos;None&apos;"/>
<Event MSec=  "9520.0975" PID="24252" PName="Process(24252)" TID="24248" EventName="Enter" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#40350611" memberName="ReceiveBlobAsync" parameters=""/>
<Event MSec=  "9520.8673" PID="24252" PName="Process(24252)" TID="24248" EventName="SecureChannelCtor" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SecureChannel#7946015" sslStream="SslStream#22286816" hostname="" clientCertificatesCount="0" encryptionPolicy="0"/>
<Event MSec=  "9520.9505" PID="24252" PName="Process(24252)" TID="24248" EventName="Enter" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#22286816" memberName="ReceiveBlobAsync" parameters=""/>
<Event MSec=  "9521.5844" PID="24252" PName="Process(24252)" TID="24248" EventName="Info" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#22286816" memberName="ReceiveBlobAsync" message="SinceSSL3"/>
<Event MSec=  "9521.6208" PID="24252" PName="Process(24252)" TID="24248" EventName="Info" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#22286816" memberName="ReceiveBlobAsync" message="frameSize=339 activeLength=339"/>
<Event MSec=  "9521.7260" PID="24252" PName="Process(24252)" TID="24248" EventName="ReceivedFrame" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#22286816" frameInfo="Tls:ClientHello[334] TargetName=&apos;testservereku.contoso.com&apos; SupportedVersion=&apos;Tls, Tls12&apos; ApplicationProtocols=&apos;None&apos;"/>
<Event MSec=  "9531.2886" PID="24252" PName="Process(24252)" TID="24248" EventName="SentFrame" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#22286816" frameInfo="Tls11:ServerHello[66] SupportedVersion=&apos;Tls11&apos; ApplicationProtocols=&apos;None&apos;"/>
<Event MSec=  "9531.3009" PID="24252" PName="Process(24252)" TID="24248" EventName="Enter" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#22286816" memberName="ReceiveBlobAsync" parameters=""/>
<Event MSec=  "9537.0314" PID="24252" PName="Process(24252)" TID="24281" EventName="Info" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#40350611" memberName="ReceiveBlobAsync" message="SinceSSL3"/>
<Event MSec=  "9537.0479" PID="24252" PName="Process(24252)" TID="24281" EventName="Info" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#40350611" memberName="ReceiveBlobAsync" message="frameSize=71 activeLength=2102"/>
<Event MSec=  "9537.0646" PID="24252" PName="Process(24252)" TID="24281" EventName="ReceivedFrame" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#40350611" frameInfo="Tls11:ServerHello[66] SupportedVersion=&apos;Tls11&apos; ApplicationProtocols=&apos;None&apos;"/>
<Event MSec=  "9537.2936" PID="24252" PName="Process(24252)" TID="24281" EventName="ErrorMessage" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#40350611" memberName="ForceAuthenticationAsync" message="InternalError"/>
<Event MSec=  "9537.4305" PID="24252" PName="Process(24252)" TID="24281" EventName="ErrorMessage" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#40350611" memberName="ForceAuthenticationAsync" message="ErrorCode=InternalError, Exception=Interop+OpenSsl+SslException: SSL Handshake failed with OpenSSL error - SSL_ERROR_SSL.
<Event MSec=  "9537.5804" PID="24252" PName="Process(24252)" TID="24281" EventName="SentFrame" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#40350611" frameInfo="Tls11:Alert[2]"/>
<Event MSec= "12189.9537" PID="24252" PName="Process(24252)" TID="24249" EventName="Enter" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#40350611" memberName="CloseInternal" parameters=""/>
<Event MSec= "12190.1188" PID="24252" PName="Process(24252)" TID="24249" EventName="Exit" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#40350611" memberName="CloseInternal" result=""/>
<Event MSec= "12190.2762" PID="24252" PName="Process(24252)" TID="24271" EventName="Info" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#22286816" memberName="ReceiveBlobAsync" message="frameSize=7 activeLength=7"/>
<Event MSec= "12190.5177" PID="24252" PName="Process(24252)" TID="24271" EventName="ErrorMessage" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#22286816" memberName="ForceAuthenticationAsync" message="InternalError"/>
<Event MSec= "12190.5354" PID="24252" PName="Process(24252)" TID="24249" EventName="Enter" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#22286816" memberName="CloseInternal" parameters=""/>
<Event MSec= "12190.6161" PID="24252" PName="Process(24252)" TID="24271" EventName="ErrorMessage" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#22286816" memberName="ForceAuthenticationAsync" message="ErrorCode=InternalError, Exception=Interop+OpenSsl+SslException: SSL Handshake failed with OpenSSL error - SSL_ERROR_SSL.
<Event MSec= "12190.7479" PID="24252" PName="Process(24252)" TID="24249" EventName="Exit" ProviderName="Microsoft-System-Net-Security" thisOrContextObject="SslStream#22286816" memberName="CloseInternal" result="”/>

```

**Comment by wfurt** [comment]
closed by  #36718
